zip-0308.rst

```diff
@@ -0,0 +1,326 @@
+increasing the latency of some other shielded operations.
+
+It is possible to enable or disable migration at any time.
+
+All shielded funds in Sprout z-addresses will eventually be transferred to
+Sapling z-addresses, provided the node is working.
+
+It should take a "reasonable" length of time to complete the transfer;
+typically less than 2 months.
+
+The design should mitigate information leakage via timing information and
+transaction data, including
+
+* linkage of particular z-addresses or users, and the amounts held;
+* information about the distribution of amounts of individual notes.
+
+The design and implementation is stateless, to the extent practical.
+
+Visibility is provided for the wallet/node user into the progress of the
+migration.
+
+There is sufficient information available to debug failed transactions that
+are part of the migration.
+
+The design recovers from failed operations to the extent possible.
+
+Each user sends a number of migration transactions that is not too small (to
+obscure their total amount).
+
+
+Non-requirements
+================
+
+There is no requirement or assumption of network layer anonymity. (Users may,
+but are not expected to, configure Tor.)
+
+The migration procedure does not have to provably leak no information.
+
+There is no need to preserve individual note values (i.e. notes can be
+consolidated).
+
+Migration txns need only be hidden among themselves, rather than among all
+kinds of transaction.
+
+A small amount (less than 0.01 ZEC) can be left unmigrated if this helps with
+privacy.
+
+
+Specification
+=============
+
+There are two main aspects to a strategy for selecting migration transactions:
+
+* how many transactions are sent, and when;
+* the amount sent in each transaction.
+
+
+Transaction schedule
+--------------------
+
+When migration is enabled, a node will send up to 5 transactions for inclusion
+in each block with height a multiple of 500 (that is, they are sent immediately
+after seeing a block with height 499, modulo 500).
+
+The migration transactions to be sent in a particular batch can take
+significant time to generate, and this time depends on the speed of the user's
+computer. If they were generated only after a block is seen at the target
+height minus 1, then this could leak information. Therefore, for target
+height N, implementations SHOULD start generating the transactions at around
+height N-5. Each migration transaction SHOULD specify an anchor at height N-10
+for each Sprout JoinSplit description.
+
+Open questions:
+
+* does this reliably give sufficient time to generate the transactions?
+* what happens to a batch if the anchor is invalidated -- should it be
+  regenerated, or cancelled?
+
+Rationale
+'''''''''
+
+Privacy is increased when the times at which to send transactions are
+coordinated between nodes. We choose to send a batch of transactions at each
+coordinated time. Sending multiple transactions in each batch ensures that:
```

Comment:
One alternative is a randomized approach, where each node sends at random times (specifically: random block-depth delays drawn from an exponential distribution).
The proposed fixed-depth approach leaks exactly whether the node's migration progress was started in time for the (n*500-5)-th block, and moreover risks congestion at the multiple-of-500 blocks. The randomized approach lessen the start-time leakage and avoid the congestion.

Reply:
I have no good intuition about this.

Reply:
It's very unlikely that there will be enough migration transactions to cause a mempool size overflow, I think.

Reply:
The randomized approach is more elegant (it's stateless and temporally invariant), and offers slightly better privacy (because for the aforementioned timing uncertainty).

The 500-blocks version has the advantage of simplicity and predictability.

Overall I would weakly prefer the randomized version, but it's OK to keep the 500-block if need be.

Reply:
In another comment I mention that we might want a random distribution after a trigger height (a third option that draws from the above two). But I'm inclined to support whichever option ends up being the simplest to implement.

Reply:
I think we should keep it as it is, for simplicity.